File: java/41.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        Map<Integer, Integer> sortedMap = new TreeMap<Integer, Integer>();
        Map<Integer, Boolean> visited = new HashMap<Integer, Boolean>();
        sortedMap.put(0, 0);
        for (Integer num : nums) {
            if (!visited.containsKey(num)) {
                visited.put(num, true);
                if (num >= 0 && !sortedMap.containsKey(num)) {
                    if (sortedMap.containsKey(num + 1)) {
                        int value = sortedMap.get(num + 1);
                        sortedMap.remove(num + 1);
                        sortedMap.put(num, value);
                    } else {
                        sortedMap.put(num, num);
                    }
                }
            }
        }
        for (Integer start : sortedMap.keySet()) {
            Integer end = sortedMap.get(start);
            if (!sortedMap.containsKey(end + 1)) {
                return end + 1;
            }
        }
        return 0;
    }
}

File: cpp/41.cc
class Solution {
    public:
        int firstMissingPositive(vector<int>& nums) {
            int length = nums.size();
            bool isOneExist = false;
            for (int i = 0; i < length; i++) {
                if (nums[i] == 1) {
                    isOneExist = true;
                } else if (nums[i] < 1 || nums[i] > length) {
                    nums[i] = 1;
                }
            }
            if (isOneExist == false) {
                return 1;
            }
            for (int i = 0; i < length; i++) {
                if (abs(nums[i]) > 1) {
                    int num = abs(nums[i]);
                    nums[num - 1] = -abs(nums[num - 1]);
                }
            }
            for (int i = 1; i < length; i++) {
                if (nums[i] > 0) {
                    return i + 1;
                }
            }
            return length + 1;
        }
    };